README.md
[![bitHound Overall Score](https://www.bithound.io/github/mtr/angular-iscroll/badges/score.svg)](https://www.bithound.io/github/mtr/angular-iscroll)
[![bitHound Dependencies](https://www.bithound.io/github/mtr/angular-iscroll/badges/dependencies.svg)](https://www.bithound.io/github/mtr/angular-iscroll/master/dependencies/npm)
[![bitHound Dev Dependencies](https://www.bithound.io/github/mtr/angular-iscroll/badges/devDependencies.svg)](https://www.bithound.io/github/mtr/angular-iscroll/master/dependencies/npm)
[![bitHound Code](https://www.bithound.io/github/mtr/angular-iscroll/badges/code.svg)](https://www.bithound.io/github/mtr/angular-iscroll)

# angular-iscroll
AngularJS module that enables iScroll 5 functionality, wrapping it in an easy-to-use directive

## Install

Install the [angular-iscroll NPM package](https://www.npmjs.com/package/angular-iscroll)
```bash
npm install --save angular-iscroll
```

Install through Bower
```bash
bower install --save angular-iscroll
```

Or, to check out a development version, start by cloning the repository, by
```bash
git clone git@github.com:mtr/angular-iscroll.git
```
Then, install the necessary dependencies:
```bash
cd angular-iscroll/
npm install
```
After that, you should have a `dist` directory with a subdirectory named `lib`:
```
dist/
└── lib
    ├── angular-iscroll.js
    └── angular-iscroll.min.js
```

### Build

To rebuild the library, run
```bash
gulp            # or "gulp watch" (to rebuild on every file change)
```

To build the examples, run
```bash
gulp examples   # (will rebuild on every file change)
```


## Demo
You may have a look at [core-layout](http://mtr.github.io/core-layout/examples/) ([GitHub repo](https://github.com/mtr/core-layout/)), an Angular demo app that shows how you can use the `iscroll` directive in a responsive-design web-app with support for both drawers (slide-out menus) and modals.  For example, the demo shows how to handle DOM content generated dynamically through [ngRepeat](https://docs.angularjs.org/api/ng/directive/ngRepeat).

## Usage

In the following, `IScroll` (with capital 'I' and 'S') refers to instances 
of the [iScroll Javascript library](http://iscrolljs.com/) that this package provides an AngularJS wrapper for. 

The main usage pattern for `angular-iscroll` is to define a dependency on the `angular-iscroll` module in your AngularJS app.  For example: 
```js
angular.module('myApp', ['angular-iscroll']);
```
or, in a Browserify-based code base:
```js
angular.module('myApp', [require('angular-iscroll').name]);
```

The `angular-iscroll` module includes both a directive, `iscroll`, and a service, `iScrollService`, which gives you access to and control over a shared, global state of whether to enable, disable, or refresh the `IScroll` instances for each `iscroll` directive instance.

Next, to use the directive, you should set up your HTML template like
```html
…
<body ng-controller="MyAppController as app"
      ng-class="{
      'iscroll-on': app.iScrollState.useIScroll,
      'iscroll-off': !app.iScrollState.useIScroll
      }">
<div class="iscroll-wrapper" iscroll>
  <div class="iscroll-scroller">
  </div>
</div>
…
```
Let me explain the essential parts of that HTML example.  First of all, the `iscroll` directive is an attribute of an element belonging to the `iscroll-wrapper` class, which wraps an element of the `iscroll-scroller` class.  Those two classes are defined in the [SASS](http://sass-lang.com/) file [dist/lib/scss/_iscroll.scss](dist/lib/scss/_iscroll.scss), but they don't have any meaning unless they occur inside an `iscroll-on` class; and that's where the shared, global state from iScrollService comes in.  The controller, `MyAppController`, in the above example exposes the state variable shared by iScrollService in its scope
```js
function MyAppController(iScrollService) {
    var vm = this;  // Use 'controller as' syntax 

    vm.iScrollState = iScrollService.state;
}
```
thereby providing a way to globally change the meaning of the `iscroll-wrapper` + `iscroll-scroller` combination.  Please note: To get more info about the "controller as" syntax, you might enjoy [John Papa's AngularJS Style Guide](https://github.com/johnpapa/angularjs-styleguide#controlleras-with-vm).

Furthermore, the global iScroll state exposed by the service should be changed through the service's `enable([signalOnly])`, `disable([signalOnly])`, and `toggle([signalOnly])` methods, where each method will change the state accordingly, and then emit a corresponding signal from `$rootScope` that gets picked up and handled by the available `angular-iscroll` directive instances.  If the `signalOnly` flag is `true`, then the state is not changed by the service method, but the signal is sent nonetheless.  If the directives receive an `iscroll:disabled` signal, they will destroy any existing `IScroll` instances, and if they receive an `iscroll:enabled` signal, they will create a new `IScroll` instances per directive instance if it doesn't already exist.

It should also be noted that during instantiation, in the directive's post-link phase, the `iscroll` directive will check the `iScrollService`'s `useIScroll` state to decide whether or not it will create an actual `IScroll` instance.  Consequently, if you would like to create an AngularJS solution that uses iScroll only on, for example, iOS devices, you should determine the current browser type early, probably inside the app controller's [configuration block](https://docs.angularjs.org/guide/module#module-loading-dependencies), and set the service's `useIscroll` state accordingly.  Please note that `angular-iscroll` does not contain any code to detect which browser or platform it is currently running on, which is a separate, complex task better solved by specialized libraries, like [platform.js](https://github.com/bestiejs/platform.js).


### Manual Interaction with Each Directive's IScroll Instance
If you want access to a scope's `IScroll` instance, you can supply an optional 
`iscroll-instance` attribute when applying the `iscroll` directive, like
```html
…
<div class="iscroll-wrapper" iscroll iscroll-instance="instance">
  <div class="iscroll-scroller">
  </div>
</div>
…
```
That way, the scope's `instance` variable will hold a reference to the actual
 `IScroll` instance, so you can access the IScroll instance's own API, for 
 example to define [custom events](http://iscrolljs.com/#custom-events) or 
 access its [scroller info](http://iscrolljs.com/#scroller-info).


### Configuration
I've designed this module so that it should be easy to configure.  First of all, you can supply per-instance options, both for `IScroll` and the directive itself, when you apply the directive.  For example
```html
<div iscroll="{mouseWheel: true, momentum: true, refreshInterval: 500}">…</div>
```
would pass along the options `{mouseWheel: true, momentum: true}` to `IScroll`, while the directive-specific configuration parameter, `{refreshInterval: 500}`, is only interpreted by the directive.  Any config option not recognized as a directive-specific option, will be forwarded to `IScroll`.

There are lots of configuration options for IScroll itself; those are best [documented by iScroll](http://iscrolljs.com/#configuring).

#### Directive Options
The directive provides two configuration options:

- `asyncRefreshDelay` (default `0`): defines the delay, in ms, before the directive asynchronously performs an IScroll.refresh().  If `false`, then no async refresh is performed.  This can come in handy when you need to wait for the DOM to be rendered before `IScroll` can know the size of its scrolling area.
- `refreshInterval` (default `false`): a delay, in ms, between each periodic iScroll.refresh().  If `false`, then no periodic refresh is performed.  This functionality can be handy in complex applications, where it might be difficult to decide when `iScrollService.refresh()` should be called, and a periodic call to `IScroll.refresh()`, for example every 500 ms, might provide a smooth user experience.  To avoid scroll stuttering caused by calls to refresh during an ongoing scroll operation, the `angular-iscroll` directive prevents `refresh()` calls if IScroll is currently performing a scroll operation.
- `invokeApply` (default `false`, since version _2.0.0_): whether or not to invoke AngularJS' `$apply()` (and thereby `$digest()`) cycle on every refresh, as determined by `asyncRefreshDelay` or `refreshInterval`.  When `false`, it will not invoke model dirty checking on every call to `IScroll.refresh()`.  This can result in huge performance gain if `refreshInterval` is set to a low value (for example 500 ms).  Example usage:

        iScrollServiceProvider.configureDefaults({
            iscroll: {
               invokeApply:false
           }
        });

    To test it, you can paste this code to your app `run` block:

        /**
         * This code measures `$digest()` performance
         * by logging digest times to the console.
         */
        var $oldDigest = $rootScope.$digest;
        var $newDigest = function() {
                console.time("$digest");
               $oldDigest.apply($rootScope);
                console.timeEnd("$digest");
        };
        $rootScope.$digest = $newDigest;

    With `invokeApply = true` and `refreshInterval = 500` you'll see that digest is run every 500ms.
    With `invokeApply = false` and `refreshInterval = 500` you'll see that digest is not invoked by `angular-iscroll`.

    Thanks to [DinkoMiletic](https://github.com/DinkoMiletic) for implementing this optimization.


#### Globally Configuring the Directive's Default Options

The `iscroll` directive gets its default configuration from the `iScrollService`.  To provide a way to easily, globally configure the defaults for all `iscroll` instances, the module defines an `iScrollServiceProvider` which can be injected into the app controller's configuration block which is guaranteed to run before the controller is used anywhere.  For example:
```js
/* @ngInject */
function _config(iScrollServiceProvider) {
    // Supply a default configuration object, eg:
    iScrollServiceProvider.configureDefaults({
        iScroll: {
            // Passed through to the iScroll library
            scrollbars: true,
            fadeScrollbars: true
        },
        directive: {
            // Interpreted by the directive
            refreshInterval: 500
        }
    });
}

angular
    .module('myApp', ['angular-iscroll'])
    .config(_config);
```
The configuration you provide this way will serve as the updated global default for all `iscroll` directive instances.

Please note that the above example relies on [ng-annotate](https://www.npmjs.com/package/ng-annotate) for adding AngularJS dependency-injection annotations during builds, as indicated by the `/* @ngInject */` comment.

## Support ##
Thanks to a generous “free for Open Source” sponsorship from [BrowserStack](https://www.browserstack.com) I've been able to test [core-layout](http://mtr.github.io/core-layout/examples/), and thereby [angular-iscroll](https://github.com/mtr/angular-iscroll/), with a plethora of devices and browsers.  The following browsers and devices has been tested successfully:

* Desktop
  * Chrome 16–48 on OS X 10.10
  * Chrome 48 on Ubuntu 14.04
  * Firefox 6.0–43.0 on OS X 10.10
  * Firefox 43.0.4 on Ubuntu 14.04
  * Internet Explorer 9, 10, and 11 on Windows 7
  * Internet Explorer Edge 12 on Windows 10
  * Opera 12.12–34 on OS X 10.10
  * Safari 5.1.7 on Windows 10
  * Safari 8.0.8 on OS X 10.10.5
  * Yandex 14.12 on OS X 10.10
* Mobile
  * Android Browser 4.0 with Android 4.0.4 on Samsung Galaxy Note 10.1
  * Android Browser 4.0 with Android 4.1.2 on Samsung Galaxy S3
  * Chrome Mobile 45 with Android 5.0.2 on Samsung Galaxy S6
  * Chrome Mobile 45 with Android 4.4 on Google Nexus 5
  * Safari 4.0.5 with iOS 4.0.1 on iPhone 4
  * Safari 5.0.2 with iOS 4.3.2 on iPad 2
  * Safari 5.1 with iOS 5.0 on iPad 2
  * Safari 5.1 with iOS 5.1 on iPhone 4S
  * Safari 6.0 with iOS 6.0 on iPhone 5
  * Safari 7.0 with iOS 7.0.4 on iPad Air
  * Safari 8.0 with iOS 8.1.1 on iPad Air 2
  * Safari 9.0 with iOS 9.0.2 on iPhone 6S
  * IE Mobile 11.0 with Windows Phone 8.1 on Nokia Lumia 520, 925, and 930

### Incompatible Browsers ###
During testing, the [core-layout](http://mtr.github.io/core-layout/examples/) demo broke in the following browsers:

* Firefox 3.6, 4, and 5 on OS X 10.10
* Internet Explorer 8 on Windows 7 (fails during jQuery version 2.2.0 initialization):
  ```
  // Use the handy event callback
  document.addEventListener( "DOMContentLoaded", completed );
  ```
  Error message: `Object doesn't support this property or method`.
* Safari 3 with iOS 3 on iPhone 3GS

This does not necessarily mean that `angular-iscroll` itself breaks in the same browsers, but the demo code did.


CHANGELOG.md
# Change Log

## [v3.6.0](https://github.com/mtr/angular-iscroll/tree/v3.6.0) (2018-01-20)
[Full Changelog](https://github.com/mtr/angular-iscroll/compare/v3.5.1...v3.6.0)

**Implemented enhancements:**

- Automated detection is too strict on Android devices. [\#41](https://github.com/mtr/angular-iscroll/issues/41)

**Fixed bugs:**

- Automated detection is too strict on Android devices. [\#41](https://github.com/mtr/angular-iscroll/issues/41)

**Closed issues:**

- FF and IE directive service and scope reference problems. [\#15](https://github.com/mtr/angular-iscroll/issues/15)

## [v3.5.1](https://github.com/mtr/angular-iscroll/tree/v3.5.1) (2017-06-13)
[Full Changelog](https://github.com/mtr/angular-iscroll/compare/v3.5.0...v3.5.1)

## [v3.5.0](https://github.com/mtr/angular-iscroll/tree/v3.5.0) (2017-06-13)
[Full Changelog](https://github.com/mtr/angular-iscroll/compare/v3.4.1...v3.5.0)

## [v3.4.1](https://github.com/mtr/angular-iscroll/tree/v3.4.1) (2017-06-13)
[Full Changelog](https://github.com/mtr/angular-iscroll/compare/v3.4.0...v3.4.1)

**Closed issues:**

- Why does it always appear "scroll-off" [\#35](https://github.com/mtr/angular-iscroll/issues/35)
- Keep getting "TypeError: IScroll is not a function" [\#16](https://github.com/mtr/angular-iscroll/issues/16)

## [v3.4.0](https://github.com/mtr/angular-iscroll/tree/v3.4.0) (2016-09-02)
[Full Changelog](https://github.com/mtr/angular-iscroll/compare/v3.2.0...v3.4.0)

**Implemented enhancements:**

- Refine the need-iScroll auto-detection rules [\#28](https://github.com/mtr/angular-iscroll/issues/28)

**Merged pull requests:**

- Fix. Fixed issue with overriding instance config with default options. [\#34](https://github.com/mtr/angular-iscroll/pull/34) ([egor-manjula](https://github.com/egor-manjula))

## [v3.2.0](https://github.com/mtr/angular-iscroll/tree/v3.2.0) (2016-01-27)
[Full Changelog](https://github.com/mtr/angular-iscroll/compare/v3.1.0...v3.2.0)

**Merged pull requests:**

- \#25 Add platform.js requirement to bower.json [\#27](https://github.com/mtr/angular-iscroll/pull/27) ([simison](https://github.com/simison))

## [v3.1.0](https://github.com/mtr/angular-iscroll/tree/v3.1.0) (2016-01-22)
[Full Changelog](https://github.com/mtr/angular-iscroll/compare/v3.0.0...v3.1.0)

## [v3.0.0](https://github.com/mtr/angular-iscroll/tree/v3.0.0) (2016-01-22)
[Full Changelog](https://github.com/mtr/angular-iscroll/compare/v2.0.1...v3.0.0)

**Implemented enhancements:**

- Automatically detect the need for using iScroll. [\#25](https://github.com/mtr/angular-iscroll/issues/25)
- Support being initially disabled. [\#24](https://github.com/mtr/angular-iscroll/issues/24)
- Address jshint issues [\#23](https://github.com/mtr/angular-iscroll/issues/23)
- Remove unmaintained example code because core-layout is well maintained [\#22](https://github.com/mtr/angular-iscroll/issues/22)
- Add bitHound configuration. [\#21](https://github.com/mtr/angular-iscroll/issues/21)
- Address npm dependency issues found in master branch \[Opened by bitHound\] [\#20](https://github.com/mtr/angular-iscroll/issues/20)

**Fixed bugs:**

- iScrollServiceProvider.configureDefaults\(\) does not perform a deep merge of options with defaultOptions. [\#26](https://github.com/mtr/angular-iscroll/issues/26)

## [v2.0.1](https://github.com/mtr/angular-iscroll/tree/v2.0.1) (2015-12-17)
[Full Changelog](https://github.com/mtr/angular-iscroll/compare/v2.0.0...v2.0.1)

## [v2.0.0](https://github.com/mtr/angular-iscroll/tree/v2.0.0) (2015-12-16)
[Full Changelog](https://github.com/mtr/angular-iscroll/compare/v1.3.3...v2.0.0)

**Merged pull requests:**

- Add option to skip digest cycle when refresh option is turned on [\#19](https://github.com/mtr/angular-iscroll/pull/19) ([DinkoMiletic](https://github.com/DinkoMiletic))
- Fixing bug where the argument passed to configureDefaults\(\) clobbers the built-in defaults [\#17](https://github.com/mtr/angular-iscroll/pull/17) ([mikechamberlain](https://github.com/mikechamberlain))

## [v1.3.3](https://github.com/mtr/angular-iscroll/tree/v1.3.3) (2015-06-07)
[Full Changelog](https://github.com/mtr/angular-iscroll/compare/v1.3.2...v1.3.3)

**Implemented enhancements:**

- Create simple configuration interface for defining event handlers. [\#10](https://github.com/mtr/angular-iscroll/issues/10)

**Merged pull requests:**

- Adds bower.json and bower install instructions [\#12](https://github.com/mtr/angular-iscroll/pull/12) ([brechtpm](https://github.com/brechtpm))

## [v1.3.2](https://github.com/mtr/angular-iscroll/tree/v1.3.2) (2015-05-23)
[Full Changelog](https://github.com/mtr/angular-iscroll/compare/v1.3.1...v1.3.2)

## [v1.3.1](https://github.com/mtr/angular-iscroll/tree/v1.3.1) (2015-05-23)
[Full Changelog](https://github.com/mtr/angular-iscroll/compare/v1.3.0...v1.3.1)

## [v1.3.0](https://github.com/mtr/angular-iscroll/tree/v1.3.0) (2015-05-23)
[Full Changelog](https://github.com/mtr/angular-iscroll/compare/v1.2.5...v1.3.0)

## [v1.2.5](https://github.com/mtr/angular-iscroll/tree/v1.2.5) (2015-05-23)
[Full Changelog](https://github.com/mtr/angular-iscroll/compare/v1.2.4...v1.2.5)

**Merged pull requests:**

- Add a Bitdeli Badge to README [\#9](https://github.com/mtr/angular-iscroll/pull/9) ([bitdeli-chef](https://github.com/bitdeli-chef))

## [v1.2.4](https://github.com/mtr/angular-iscroll/tree/v1.2.4) (2015-03-25)
[Full Changelog](https://github.com/mtr/angular-iscroll/compare/v1.2.3...v1.2.4)

**Fixed bugs:**

- Interval is not being closed after destroying instance of angular-iscroll [\#6](https://github.com/mtr/angular-iscroll/issues/6)

**Merged pull requests:**

- Cancelling intervals when destroying instance [\#7](https://github.com/mtr/angular-iscroll/pull/7) ([wkwiatek](https://github.com/wkwiatek))

## [v1.2.3](https://github.com/mtr/angular-iscroll/tree/v1.2.3) (2015-03-05)
[Full Changelog](https://github.com/mtr/angular-iscroll/compare/v1.2.2...v1.2.3)


lib/angular-iscroll.js
/**
 * @license angular-iscroll v3.6.1, 2018-01-21T00:21:01+0100
 * (c) 2018 Martin Thorsen Ranang <mtr@ranang.org>
 * License: MIT
 */
(function (root, factory) {
    // Using the Universal Module Definition pattern from
    // https://github.com/umdjs/umd/blob/master/returnExports.js
    if (typeof define === 'function' && define.amd) {
        define(['iscroll', 'platform'], factory);
    } else if (typeof exports === 'object') {
        module.exports = factory(require('iscroll'), require('platform'));
    } else {
        // Browser globals (root is window)
        root.angularIscroll = factory(root.IScroll, root.platform);
    }
}(this, function (IScroll, platform) {
    'use strict';

    iscroll.$inject = ["$rootScope", "$timeout", "$interval", "iScrollSignals", "iScrollService"];
    var signals = {
            disabled: 'iscroll:disabled',
            enabled: 'iscroll:enabled',
            refresh: 'iscroll:refresh'
        },
        classes = {
            on: 'iscroll-on',
            off: 'iscroll-off'
        },
        iScrollEvents = [
            'beforeScrollStart',
            'scrollCancel',
            'scrollStart',
            'scroll',
            'scrollEnd',
            'flick',
            'zoomStart',
            'zoomEnd'
        ],
        iScrollEventHandlerMap = {},
        useNativeScroll = angular.isDefined(platform) && _useNativeScroll(platform);

    /**
     * Compares two software version numbers (e.g. "1.7.1" or "1.2b").
     *
     * This function was born in http://stackoverflow.com/a/6832721.
     *
     * @param {string} v1 The first version to be compared.
     * @param {string} v2 The second version to be compared.
     * @param {object} [options] Optional flags that affect comparison behavior:
     * <ul>
     *     <li>
     *         <tt>lexicographical: true</tt> compares each part of the version strings lexicographically instead of
     *         naturally; this allows suffixes such as "b" or "dev" but will cause "1.10" to be considered smaller than
     *         "1.2".
     *     </li>
     *     <li>
     *         <tt>zeroExtend: true</tt> changes the result if one version string has less parts than the other. In
     *         this case the shorter string will be padded with "zero" parts instead of being considered smaller.
     *     </li>
     * </ul>
     * @returns {number|NaN}
     * <ul>
     *    <li>0 if the versions are equal</li>
     *    <li>a negative integer iff v1 < v2</li>
     *    <li>a positive integer iff v1 > v2</li>
     *    <li>NaN if either version string is in the wrong format</li>
     * </ul>
     *
     * @copyright by Jon Papaioannou (["john", "papaioannou"].join(".") + "@gmail.com")
     * @license This function is in the public domain. Do what you want with it, no strings attached.
     */
    function versionCompare(v1, v2, options) {
        var lexicographical = options && options.lexicographical,
            zeroExtend = options && options.zeroExtend,
            v1parts = v1.split('.'),
            v2parts = v2.split('.');

        function isValidPart(x) {
            return (lexicographical ? /^\d+[A-Za-z]*$/ : /^\d+$/).test(x);
        }

        if (!v1parts.every(isValidPart) || !v2parts.every(isValidPart)) {
            return NaN;
        }

        if (zeroExtend) {
            while (v1parts.length < v2parts.length) v1parts.push('0');
            while (v2parts.length < v1parts.length) v2parts.push('0');
        }

        if (!lexicographical) {
            v1parts = v1parts.map(Number);
            v2parts = v2parts.map(Number);
        }

        for (var i = 0; i < v1parts.length; ++i) {
            if (v2parts.length == i) {
                return 1;
            }

            if (v1parts[i] == v2parts[i]) {
                continue;
            }
            else if (v1parts[i] > v2parts[i]) {
                return 1;
            }
            else {
                return -1;
            }
        }

        if (v1parts.length != v2parts.length) {
            return -1;
        }

        return 0;
    }

    function _isChromeMobile(platform) {
        return platform.name === 'Chrome Mobile';
    }

    function _isRecentOS(platform) {
        return versionCompare(platform.os.version, '4.0.4') >= 0;
    }

    function _useNativeScroll(platform) {
        if (platform.name === 'Opera Mini') {
            return false;
        }

        if (platform.name === 'IE Mobile') {
            return versionCompare(platform.version, '11.0') >= 0
        }

        switch (platform.os.family) {
            case 'Android':
                // In Chrome we trust, but also younger Android releases.
                return _isChromeMobile(platform) || _isRecentOS(platform);
            case 'iOS':
                // Buggy handling in older iOS versions.
                return versionCompare(platform.version, '5.1') >= 0;
            default:
                // Assuming desktop or other browser.
                return true;
        }
    }

    /**
     * Add handler name to event name mapping.
     *
     * Please note that the 'scroll' event is only available when using
     * iscroll-probe.js (for example, through angular-iscroll-probe).
     *
     * For example, the handler for the 'scrollEnd' event can be configured
     * by supplying the onScrollEnd option.
     **/
    angular.forEach(iScrollEvents, function _addPair(event) {
        this['on' + _capitalizeFirst(event)] = event;
    }, iScrollEventHandlerMap);

    function _capitalizeFirst(str) {
        return str.substring(0, 1).toLocaleUpperCase() + str.substring(1);
    }

    function iScrollServiceProvider() {
        iScrollService.$inject = ["$rootScope", "iScrollSignals"];
        var defaultOptions = {
            iScroll: {
                /**
                 * The different options for iScroll are explained in
                 * detail at http://iscrolljs.com/#configuring
                 **/
                momentum: true,
                mouseWheel: true
            },
            directive: {
                /**
                 * Whether or not to initially enable the use of iScroll.
                 *
                 * The `useNativeScroll` flag is automatically determined
                 * by running _useNativeScroll();
                 **/
                initiallyEnabled: !useNativeScroll,
                /**
                 * Delay, in ms, before we asynchronously perform an
                 * iScroll.refresh().  If false, then no async refresh is
                 * performed.
                 **/
                asyncRefreshDelay: 0,
                /**
                 * Delay, in ms, between each iScroll.refresh().  If false,
                 * then no periodic refresh is performed.
                 **/
                refreshInterval: false,
                /**
                 * If `false`, skip `$digest()` cycle on iScroll.refresh().
                 */
                invokeApply: false
                /**
                 * Event handler options are added below.
                 **/
            }
        };

        angular.forEach(iScrollEventHandlerMap, function _default(event, handler) {
            this[handler] = undefined;
        }, defaultOptions.directive);

        function _configureDefaults(options) {
            /**
             * Since angular.extend is not performing a "deep" merge, we'll
             * do it in two steps.
             **/
            if (angular.isDefined(options.directive)) {
                angular.extend(defaultOptions.directive, options.directive);
            }
            if (angular.isDefined(options.iScroll)) {
                angular.extend(defaultOptions.iScroll, options.iScroll);
            }
        }

        // Export the auto-determined value of `useNativeScroll`.
        this.useNativeScroll = useNativeScroll;
        this.platform = platform;

        this.configureDefaults = _configureDefaults;

        function _getDefaults() {
            return defaultOptions;
        }

        //noinspection JSUnusedGlobalSymbols
        this.getDefaults = _getDefaults;

        /* @ngInject */
        function iScrollService($rootScope, iScrollSignals) {
            var _state = {
                useIScroll: defaultOptions.directive.initiallyEnabled,
                autoDetectedUseNativeScroll: useNativeScroll
            };

            function _disable(signalOnly) {
                if (!signalOnly) {
                    _state.useIScroll = false;
                }
                //$log.debug('emit(iScrollSignals.disabled)',
                //    iScrollSignals.disabled);
                $rootScope.$emit(iScrollSignals.disabled);
            }

            function _enable(signalOnly) {
                if (!signalOnly) {
                    _state.useIScroll = true;
                }
                //$log.debug('emit(iScrollSignals.enabled)',
                //    iScrollSignals.enabled);
                $rootScope.$emit(iScrollSignals.enabled);
            }

            function _toggle(signalOnly) {
                (_state.useIScroll ^ signalOnly) ?  // XOR
                    _disable(signalOnly) : _enable(signalOnly);
            }

            function _refresh(name) {
                // The name parameter is not really used for now.
                $rootScope.$emit(iScrollSignals.refresh, name);
            }

            return {
                defaults: defaultOptions,
                state: _state,
                versionCompare: versionCompare,
                platform: platform,
                enable: _enable,
                disable: _disable,
                toggle: _toggle,
                refresh: _refresh
            };
        }

        //noinspection JSUnusedGlobalSymbols
        this.$get = iScrollService;
    }

    function _call(functor) {
        functor();
    }

    /* @ngInject */
    function iscroll($rootScope, $timeout, $interval, iScrollSignals,
                     iScrollService) {
        function asyncRefresh(instance, options) {
            $timeout(function _refreshAfterInitialRender() {
                instance.refresh();
            }, options.directive.asyncRefreshDelay, options.directive.invokeApply);
        }

        function _createInstance(scope, element, attrs, options) {
            var instance = new IScroll(element[0], options.iScroll),
                refreshEnabled = true,
                refreshInterval = null;

            angular.forEach(iScrollEventHandlerMap,
                function _addHandler(event, option) {
                    if (angular.isDefined(options.directive[option])) {
                        instance.on(event, options.directive[option]);
                    }
                });

            element.removeClass(classes.off).addClass(classes.on);

            if (angular.isDefined(attrs.iscrollInstance)) {
                scope.iscrollInstance = instance;
            }

            if (options.directive.asyncRefreshDelay !== false) {
                asyncRefresh(instance, options);
            }

            function _destroyInstance() {
                if (refreshInterval !== null) {
                    $interval.cancel(refreshInterval);
                }

                if (angular.isDefined(scope.iscrollInstance)) {
                    delete scope.iscrollInstance;
                }
                instance.destroy();

                element.removeClass(classes.on).addClass(classes.off);
                // Remove element's CSS transition values:
                element.children('.iscroll-scroller').attr('style', null);

                angular.forEach(deregistrators, _call);
                //$log.debug('angular-iscroll: destroyInstance');
            }

            function _refreshInstance() {
                if (refreshEnabled) {
                    //noinspection JSUnusedAssignment
                    refreshEnabled = false;
                    asyncRefresh(instance, options);
                    refreshEnabled = true;
                }
            }

            function _disableRefresh() {
                refreshEnabled = false;
            }

            function _enableRefresh() {
                refreshEnabled = true;
            }

            instance.on('scrollStart', _disableRefresh);
            instance.on('scrollEnd', _enableRefresh);

            if (options.directive.refreshInterval !== false) {
                refreshInterval = $interval(_refreshInstance,
                    options.directive.refreshInterval, 0, options.directive.invokeApply);
            }

            var deregistrators = [
                $rootScope.$on(iScrollSignals.disabled, _destroyInstance),
                $rootScope.$on(iScrollSignals.refresh, _refreshInstance),
                scope.$on('$destroy', _destroyInstance)
            ];

            return instance;
        }

        function _link(scope, element, attrs) {
            var options = {
                iScroll: angular.extend({}, iScrollService.defaults.iScroll, scope.iscroll || {}),
                directive: angular.extend({}, iScrollService.defaults.directive)
            };

            angular.forEach(options.iScroll,
                function _extractOptions(value, key) {
                    if (iScrollService.defaults.directive.hasOwnProperty(key) ||
                        iScrollEventHandlerMap.hasOwnProperty(key)) {
                        options.directive[key] = value;
                        delete options.iScroll[key];
                    }
                }
            );

            function _init() {
                if (!element.hasClass(classes.on)) {
                    _createInstance(scope, element, attrs, options);
                }
            }

            var enableHandlers = [$rootScope.$on(iScrollSignals.enabled, _init)];

            function _removeEnableHandlers() {
                angular.forEach(enableHandlers, _call);
            }

            if (iScrollService.state.useIScroll) {
                _init();
            } else {
                element.removeClass(classes.on).addClass(classes.off);
            }

            scope.$on('$destroy', _removeEnableHandlers);
        }

        return {
            restrict: 'A',
            link: _link,
            scope: {
                iscroll: '=',
                iscrollInstance: '='
            }
        };
    }

    return angular.module('angular-iscroll', [])
        .directive('iscroll', iscroll)
        .provider('iScrollService', iScrollServiceProvider)
        .constant('iScrollSignals', signals);
}));


